core/src/consensus/stake/action_data.rs

```diff
@@ -257,6 +259,74 @@ impl IntermediateRewards {
     }
 }
 
+pub struct Candidates(BTreeMap<Address, Candidate>);
+#[derive(RlpEncodable, RlpDecodable, Clone)]
+#[cfg_attr(test, derive(Debug, Eq, PartialEq))]
+pub struct Candidate {
+    pub address: Address,
+    pub deposit: Deposit,
+    pub nomination_ends_at: u64,
+}
+
+impl Candidates {
+    pub fn load_from_state(state: &TopLevelState) -> StateResult<Candidates> {
+        let key = *CANDIDATES_KEY;
+        let candidates = state.action_data(&key)?.map(|data| decode_list::<Candidate>(&data)).unwrap_or_default();
+        let indexed = candidates.into_iter().map(|c| (c.address, c)).collect();
+        Ok(Candidates(indexed))
+    }
+
+    pub fn save_to_state(&self, state: &mut TopLevelState) -> StateResult<()> {
+        let key = *CANDIDATES_KEY;
+        if !self.0.is_empty() {
+            let encoded = encode_iter(self.0.values());
+            state.update_action_data(&key, encoded)?;
+        } else {
+            state.remove_action_data(&key);
+        }
+        Ok(())
+    }
+
+    pub fn get_candidate(&self, account: &Address) -> Option<&Candidate> {
+        self.0.get(&account)
+    }
+
+    #[cfg(test)]
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn add_deposit(&mut self, address: &Address, quantity: Deposit, nomination_ends_at: u64) {
+        match self.0.entry(*address) {
```

Comment:
How about
```rust
        let candidate = self.0.entry(*address).or_insert_with(|| {
            Candidate {
                address: *address,
                deposit: 0,
                nomination_ends_at: 0,
            }
        });
        assert!(
            candidate.nomination_ends_at <= nomination_ends_at,
            "nomination_ends_at should not be decreased (existing {}, new {})",
            candidate.nomination_ends_at,
            nomination_ends_at
        );
        candidate.deposit += quantity;
        candidate.nomination_ends_at = nomination_ends_at;
```

Reply:
I've removed the assertion to allow `nomitation_ends_at` to be decreased.